compiler/rustc_data_structures/src/sync.rs

```diff
@@ -360,6 +478,72 @@ cfg_if! {
     }
 }
 
+// Ensure data structures is `Send` by using runtime check.
+// Note: Users need to ensure that the data structure **will not break**
+// thread safety after being created.
+pub unsafe auto trait DynSend {}
+
+// Ensure data structures is `Sync` by using runtime check.
+// Note: Users need to ensure that the data structure **will not break**
+// thread safety after being checked.
+pub unsafe auto trait DynSync {}
+
+impl<T> !DynSend for *mut T {}
+impl<T> !DynSync for *mut T {}
+
+// If `T` is `Sync`, then `&T` is `Send`.
+unsafe impl<T: DynSync + ?Sized> DynSend for &T {}
+
+unsafe impl<T: Send> DynSend for std::sync::Mutex<T> {}
+unsafe impl<T: Sync> DynSync for std::sync::Mutex<T> {}
```

Comment:
Can these be `for T`?

Reply:
I think there are still big problems in the implementation of auto trait in the current language. I tried two situation:
1. negative + positive impl:
```
impl<T: ?Sized> !DynSend for *mut T {}
unsafe impl<T: ?Sized + Send> DynSend for T {}
```
Then it failed with:
```
error[E0751]: found both positive and negative implementation of trait `DynSend` for type `*mut _`:
   --> compiler\rustc_data_structures\src\sync.rs:492:1
    |
491 | impl<T: ?Sized> !DynSend for *mut T {}
    | ----------------------------------- negative implementation here
492 | unsafe impl<T: Send> DynSend for T {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ positive implementation here
```
2. all positive impl:
```
unsafe impl<T: ?Sized + Sync> DynSync for T {}
// We will use `RefCell` and `Mutex` in `Lock` in the future
unsafe impl<T> DynSync for RefCell<T> {}
```
Then it failed with:
```
error[E0119]: conflicting implementations of trait `DynSync` for type `RefCell<_>`
   --> compiler\rustc_data_structures\src\sync.rs:498:1
    |
496 | unsafe impl<T: ?Sized + Sync> DynSync for T {}
    | ------------------------------------------- first implementation here
497 | // We will use `RefCell` and `Mutex` in `Lock` in the future
498 | unsafe impl<T> DynSync for RefCell<T> {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `RefCell<_>`
```

Reply:
For the second I found a solution:
```
unsafe impl<T: ?Sized + Send> DynSend for T {}
unsafe impl<T: ?Sized + Sync> DynSync for T {}

pub struct SLock<T>(std::cell::RefCell<T>, usize);
impl<T> !Sync for SLock<T> {}
unsafe impl<T> DynSync for SLock<T> {}
```
But I encountered a problem about auto trait again, its MSVC is as follows:
```
#![feature(auto_traits)]

auto trait DynSend {}

impl<T: ?Sized + Send> DynSend for T {}

fn foo(f: impl Fn() + DynSend) {
    f()
}

fn foo2(f: impl Fn() + DynSend) {
    // `foo(f)` is ok
    foo(|| f())
}

fn main() {
    foo(|| println!("hello!"));
    foo2(|| println!("hello!"));
}
```
`DynSend` doesn't seem to be passing across closures